File: store-frontend/src/shared/services/auth.service.ts
import { ApiService } from "@/shared/services/api.service";
import { User } from "@/features/users/types/usersTypes";
import { LoginRequest } from "@/features/auth/login/types";
import { RegisterRequest } from "@/features/auth/register/types";
import { useAuthStore } from "@/store/auth.store";

export interface AuthResponse {
    user: User;
}

export const AuthService = {
    login: async (request: LoginRequest): Promise<AuthResponse> => {
        const response = await ApiService.post<AuthResponse>("/auth/login", request);
        useAuthStore.getState().setSession(response.data.user);
        return response.data;
    },

    register: async (request: RegisterRequest): Promise<AuthResponse> => {
        const response = await ApiService.post<AuthResponse>("/auth/register", request);
        useAuthStore.getState().setSession(response.data.user);
        return response.data;
    },

    logout: async (): Promise<void> => {
        try {
            await ApiService.post<void>("/auth/logout");
        } finally {
            useAuthStore.getState().clearSession();
        }
    },

    refreshToken: async (): Promise<AuthResponse> => {
        const response = await ApiService.post<AuthResponse>("/auth/refresh-token", {});
        useAuthStore.getState().setSession(response.data.user);
        return response.data;
    },

    verifyToken: async (): Promise<User> => {
        const response = await ApiService.get<User>("/auth/me");
        const user = response.data;
        useAuthStore.getState().setSession(user);
        return user;
    },
};

File: store-backend/src/main/java/com/yisus/store_backend/auth/service/AuthService.java
package com.yisus.store_backend.auth.service;

import com.yisus.store_backend.auth.dto.AuthSessionResult;
import com.yisus.store_backend.auth.dto.LoginRequest;
import com.yisus.store_backend.auth.dto.RegisterRequest;
import com.yisus.store_backend.role.model.Role;
import com.yisus.store_backend.role.repository.RoleRepository;
import com.yisus.store_backend.user.dto.UserDTO;
import com.yisus.store_backend.user.model.User;
import com.yisus.store_backend.user.repository.UserRepository;
import com.yisus.store_backend.user.service.UserService;

import lombok.RequiredArgsConstructor;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final RoleRepository roleRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;
    private final AuthenticationManager authenticationManager;
    private final UserService userService;
    private final TokenBlacklistService tokenBlacklistService;

    @Transactional
    public AuthSessionResult register(RegisterRequest request) {
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new RuntimeException("Email already registered");
        }

        Role userRole = roleRepository.findByName("USER")
                .orElseThrow(() -> new RuntimeException("USER role not found. System may not be properly initialized."));

        User user = User.builder()
                .name(request.getName())
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword()))
                .role(userRole)
                .isActive(true)
                .build();

        userRepository.save(user);

        String accessToken = jwtService.generateAccessToken(user);
        String refreshToken = jwtService.generateRefreshToken(user);

        UserDTO userDTO = userService.mapToDTO(user);

        return new AuthSessionResult(userDTO, accessToken, refreshToken);
    }

    public AuthSessionResult login(LoginRequest request) {
        authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(
                        request.getEmail(),
                        request.getPassword()
                )
        );

        User user = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new RuntimeException("User not found"));

        String accessToken = jwtService.generateAccessToken(user);
        String refreshToken = jwtService.generateRefreshToken(user);

        UserDTO userDTO = userService.mapToDTO(user);

        return new AuthSessionResult(userDTO, accessToken, refreshToken);
    }

    public AuthSessionResult refreshToken(String refreshToken) {
        if (tokenBlacklistService.isTokenBlacklisted(refreshToken)) {
            throw new BadCredentialsException("Refresh token has been revoked");
        }

        String userEmail = jwtService.extractUsername(refreshToken);

        User user = userRepository.findByEmail(userEmail)
                .orElseThrow(() -> new RuntimeException("User not found"));

        if (!jwtService.isTokenValid(refreshToken, user)) {
            throw new RuntimeException("Invalid refresh token");
        }

        tokenBlacklistService.blacklistToken(refreshToken);

        String newAccessToken = jwtService.generateAccessToken(user);
        String newRefreshToken = jwtService.generateRefreshToken(user);

        UserDTO userDTO = userService.mapToDTO(user);

        return new AuthSessionResult(userDTO, newAccessToken, newRefreshToken);
    }

    public void logout(String accessToken, String refreshToken) {
        if (accessToken != null && !accessToken.isBlank()) {
            tokenBlacklistService.blacklistToken(accessToken);
        }
        if (refreshToken != null && !refreshToken.isBlank()) {
            tokenBlacklistService.blacklistToken(refreshToken);
        }
    }

    public UserDTO getCurrentUser() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found"));
        return userService.mapToDTO(user);
    }
}
